docs/design-proposals/2022-04 Parquet.md

```diff
@@ -0,0 +1,256 @@
+                    optional binary Value (STRING);
+                    optional binary ValueArray;
+                    optional boolean ValueBool;
+                    optional double ValueDouble;
+                    optional int64 ValueInt (INT(64,true));
+                    optional binary ValueKVList;
+                }
+
+                repeated group Events {
+                    repeated group Attrs {
+                        required binary Key (STRING);
+                        required binary Value (STRING);
+                    }
+                    required int32 DroppedAttributesCount (INT(32,true));
+                    required binary Name (STRING);
+                    required int64 TimeUnixNano (INT(64,false));
+                }
+
+                # Dedicated columns for common span attributes
+                optional binary HttpMethod (STRING);
+                optional binary HttpUrl (STRING);
+                optional int64 HttpStatusCode (INT(64,true));
+            }
+
+            required group InstrumentationLibrary {
+                required binary Name (STRING);
+                required binary Version (STRING);
+            }
+        }
+    }
+}
+```
+
+### Trace-level attributes
+For speed and ease-of-use, we are projecting several values to columns at the trace-level:
+
+* Trace ID - Don't store on each span.  It is a UTF-8 hexadecimal string for user-friendliness. We could save a few bytes by storing the underlying 16 bytes, but this makes it easier to work with for downstream applications.
+* Root service/span names/StartTimeUnixNano - These are selected properties of the root span in each trace (if there is one). These are used for display of results in the Grafana UI. Computed at ingest time and stored once for efficiency, so we don't have to find the root span.
+* DurationNanos - The total trace duration, computed at ingest time. This powers the min/max duration filtering in the current Tempo search and is more efficient than scanning the spans duration column. However, it may go away with TraceQL or we could decide to change it to span-level duration filtering too.
+
+
+### Dedicated columns
+Projecting attributes to their own columns has massive benefits for speed and size, and these are too good to pass up on.  Therefore we are taking an opinionated approach and projecting some common attributes to their own columns.  All other attributes are stored in the generic key/value maps and are still searchable, but not as quickly.  We chose these attributes based on what we commonly use ourselves (scratching our own itch), but we think they will be useful to most workloads.  There is an associated cost for unused attributes as it will still store a column of nulls, however the cost should be minimal.  We can be more generous with unused resource-level attributes because their overhead is the smallest of all, we are finding it to be ~.05% (0.0005) of the total block size.
+
+Resource-level:
```

Comment:
If we made these configurable would we have to store it in the block meta somewhere? or would it be dynamically detected during querying?

I feel like it would blow up the block meta too much to store these for every block, but I like the idea of having a configurable list of "dedicated columns".

Reply:
Even if we added a configurable list of dedicated columns, the schema together with all columns would be part of the parquet metadata, and we are looking at ways to cache that to prevent pulling it from the object store on every read. So I think it will be "fine" until it blows up to a million columns, but I'm just speculating.

Reply:
Making the columns configurable is a good idea and we had the same thought to be "less opinionated", but it ends up as a dynamic schema still, which is non-trivial and presents several challenges that we were avoiding in this first pass.  (1) read/write using reflection since a static struct is not available (2) attribute name -> column name translation (3) type-handling (unless name + type is part of the configuration).  

Yes the plan for when we get to dynamic schema is to detect the columns at runtime during querying.